src/Aspire.Hosting/Utils/PeriodicRestartAsyncEnumerable.cs

```diff
@@ -0,0 +1,121 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+namespace Aspire.Hosting.Utils;
+
+internal static class PeriodicRestartAsyncEnumerable
+{
+    /// <summary>
+    /// Creates an async enumerable that wraps and periodically restarts an inner async enumeration. This is intended to keep long
+    /// running watch enumerations fresh and will recreate the watch enumeration on a set interval or if the inner enumerable terminates
+    /// unexpectedly. The goal is to ensure that we keep the wrapped enumeration active until the main token is cancelled.
+    /// </summary>
+    /// <typeparam name="T">The type the enumerable iterates over</typeparam>
+    /// <param name="enumerableFactory">Factory method that takes the last iterrated value (if one exists) and a <see cref="CancellationToken"/> and returns a fresh <see cref="IAsyncEnumerable{T}"/> to enumerate over</param>
+    /// <param name="restartInterval">How often should we get a new enumerable from the factory</param>
+    /// <param name="cancellationToken">Stop all enumeration once this is cancelled</param>
+    /// <returns>An <see cref="IAsyncEnumerable{T}"/> of items returned by the inner iterables</returns>
+    public static async IAsyncEnumerable<T> CreateAsync<T>(Func<T?, CancellationToken, Task<IAsyncEnumerable<T>>> enumerableFactory, TimeSpan restartInterval, [EnumeratorCancellation] CancellationToken cancellationToken) where T : struct
```

Comment:
Generics don't support using T? with a notnull constraint since T? is handled differently for value and reference types.